rust/kernel/kunit.rs

```diff
@@ -0,0 +1,83 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! KUnit-based macros for Rust unit tests.
+//!
+//! C header: [`include/kunit/test.h`](../../../../../include/kunit/test.h)
+//!
+//! Reference: <https://www.kernel.org/doc/html/latest/dev-tools/kunit/index.html>
+
+/// Asserts that a boolean expression is `true` at runtime.
+///
+/// Public but hidden since it should only be used from generated tests.
+///
+/// Unlike the one in `core`, this one does not panic; instead, it is mapped to the KUnit
+/// facilities. See [`assert!`] for more details.
+#[doc(hidden)]
+#[macro_export]
+macro_rules! kunit_assert {
+    ($test:expr, $cond:expr $(,)?) => {{
+        if !$cond {
+            #[repr(transparent)]
+            struct Location($crate::bindings::kunit_loc);
+
+            #[repr(transparent)]
+            struct UnaryAssert($crate::bindings::kunit_unary_assert);
+
+            // SAFETY: There is only a static instance and in that one the pointer field
+            // points to an immutable C string.
+            unsafe impl Sync for Location {}
+
+            // SAFETY: There is only a static instance and in that one the pointer field
+            // points to an immutable C string.
+            unsafe impl Sync for UnaryAssert {}
+
+            static FILE: &'static $crate::str::CStr = $crate::c_str!(core::file!());
+            static LOCATION: Location = Location($crate::bindings::kunit_loc {
+                file: FILE.as_char_ptr(),
+                line: core::line!() as i32,
+            });
+            static CONDITION: &'static $crate::str::CStr = $crate::c_str!(stringify!($cond));
+            static ASSERTION: UnaryAssert = UnaryAssert($crate::bindings::kunit_unary_assert {
+                assert: $crate::bindings::kunit_assert {
+                    format: Some($crate::bindings::kunit_unary_assert_format),
+                },
+                condition: CONDITION.as_char_ptr(),
+                expected_true: true,
+            });
+
+            // SAFETY:
+            //   - FFI call.
+            //   - The `test` pointer is valid because this hidden macro should only be called by
+            //     the generated documentation tests which forward the test pointer given by KUnit.
+            //   - The string pointers (`file` and `condition`) point to null-terminated ones.
+            //   - The function pointer (`format`) points to the proper function.
+            //   - The pointers passed will remain valid since they point to statics.
+            //   - The format string is allowed to be null.
+            unsafe {
+                $crate::bindings::kunit_do_failed_assertion(
```

Comment:
Does this function do a longjmp? If so it is UB to use if there is any stackframe with destructors on the stack between the setjmp and longjmp. Furthermore it would violate the `Pin` invariant that memory is not reused unless the `Drop` impl runs.

Reply:
To be fair, currently `panic!`king will cause the running thread/kthread to be killed, and the kernel will also reclaim the stack, leading to the same issue.

Reply:
For the Pin invariant violation, indeed. Longjmp being UB does not cause issues for killing a thread.

Reply:
@bjorn3 and I discussed this and we decided this is reasonable for the moment. He will open an issue describing several ideas/approaches we could take (thanks a lot!). For the moment, we added this comment:

```
//   - There are, however, problems with this: first of all, this will end up stopping
//     the thread, without running destructors. While that is problematic in itself,
//     it is considered UB to have what is effectively an unforced foreign unwind
//     with `extern "C"` ABI. One could observe the stack that is now gone from
//     another thread. We should avoid pinning stack variables to prevent library UB,
//     too. For the moment, given test failures are reported immediately before the
//     next test runs, that test failures should be fixed and that KUnit is explicitly
//     documented as not suitable for production environments, we feel it is reasonable.
```